include/dlaf/matrix/copy.h

```diff
@@ -13,37 +13,58 @@
 #include <pika/execution.hpp>
 #include <pika/future.hpp>
 
+#include "dlaf/common/range2d.h"
 #include "dlaf/matrix/copy_tile.h"
+#include "dlaf/matrix/matrix.h"
 #include "dlaf/types.h"
-#include "dlaf/util_matrix.h"
 
-namespace dlaf {
-namespace matrix {
+namespace dlaf::matrix {
 
-/// Copy values from another matrix.
+/// Copy subset of local tiles from @p source in the range @p idx_source_begin, @p sz to local subset of
+/// tiles in @p dest in the range @p idx_dest_begin, @p sz.
 ///
-/// Given a matrix with the same geometries and distribution, this function submits tasks that will
-/// perform the copy of each tile.
 template <class T, Device Source, Device Destination>
-void copy(Matrix<const T, Source>& source, Matrix<T, Destination>& dest) {
-  const auto& distribution = source.distribution();
-
-  DLAF_ASSERT(matrix::equal_size(source, dest), source, dest);
-  DLAF_ASSERT(matrix::equal_blocksize(source, dest), source, dest);
-  DLAF_ASSERT(matrix::equal_distributions(source, dest), source, dest);
+void copy(LocalTileSize sz, LocalTileIndex idx_source_begin, Matrix<const T, Source>& source,
+          LocalTileIndex idx_dest_begin, Matrix<T, Destination>& dest) {
+  // Note: the constexpr check is ncessary to avoid compiler errors for comparison of distinct pointer types
+  if constexpr (Source == Destination) {
+    if (idx_source_begin == idx_dest_begin && &source == &dest)
+      return;
+  }
```

Comment:
> This code made me wonder if it would be helpful to add a check for verifying that ranges in source and destination are disjoint when operating with the same matrix, otherwise it might end up with wrong data copied without any deadlock (or a deadlock). It's a kind of memory pointer aliasing.

I agree that the result may be unexpected but I don't see how a deadlock or a race condition may occur. As you have pointed out, we have had this problem with the original implementation too.

Would it be enough to leave a cautionary note in the documentation about copying overlapping 2D regions of the same matrix?

Alternatively, should the function just return doing nothing if there is any overlap (not only full overlap)?

I am hesitant to put a `DLAF_ASSERT()` for overlap because I don't want it throw an error if it's the same matrix and the same ranges and if I don't throw an error in that case, throwing an error for partial overlap seems inconsistent.

Reply:
No, you're right, it's not about deadlock, but the problem about wrong results still stands in cases as described above.

IMHO right now we can just go with opening an issue and leave it for a possible change in the future. I would not stop this PR.

@rasolca @msimberg do you agree?

Reply:
I'm indifferent. I don't know the use cases of this overload well enough to judge if this could be worth checking.

Reply:
I added a comment about partial overlap here. https://github.com/eth-cscs/DLA-Future/pull/732/commits/45606d4c3fcdc27a7c69a2b3b9ae22a3de9ad17a